cioos_data_transform/cioos_data_transform/utils/xarray.py

```diff
@@ -0,0 +1,152 @@
+                    except ValueError:
+                        warnings.warn('Failed to read date {0}: {1}. Will try to use pandas.to_datetime()'\
+                                             .format(var, value), RuntimeWarning)
+                        value = pd.to_datetime(value)
+                else:
+                    # Try with Pandas
+                    value = pd.to_datetime(value)
+            elif info['format'] == 'json':
+                value = json.dumps(value, ensure_ascii=False, indent=False)
+            else:
+                value = np.array(value).astype(info['format'])
+
+        # Add variable to xarray dataset
+        if global_attribute:
+            ds.attrs[var] = value
+        elif var in [time, latitude, longitude, depth]:
+            ds.coords[var] = value
+        else:
+            ds[var] = value
+
+    return ds
+
+
+import numpy as np
+
+
+def convert_variables_to_erddap_format(ds):
+    """
+    convert_variables_to_erddap_format converts each variables within an xarray to an
+    ERDDAP compatible/recommended format.
+      - datetime (timezone aware or not) are converted to: seconds since 1970-01-01T00:00:00[Z]
+      - Any objects (usually strings) are converted to |S
+    """
+    variables_to_review = list(ds.keys())
+    variables_to_review.extend(ds.coords.keys())
+    for var in variables_to_review:
+        if ds[var].dtype not in [float, int, 'float64', 'float32', 'int64', 'int32']:
+            # Convert Datetime to seconds since 1970-01-01
+            if ds[var].dtype.name.startswith('datetime'):
+                # Convert Datetime to seconds since 1970-01-01
+                if 'units' in ds[var].attrs:
+                    ds[var].attrs.pop('units')
+                # Timezone aware data
+                if 'tz' in ds[var].dtype.name:
+                    timezone = 'Z'
+                else:
+                    timezone = ''
+
+                # Format encoding output
+                ds[var].encoding['units'] = 'seconds since 1970-01-01 00:00:00' + timezone
+
+            else:
+                # Should be a string
+                ds[var] = ds[var].astype(str).str.encode('utf-8').astype("|S")
+                ds[var].attrs['_Encoding'] = 'UTF-8'
+    return ds
+
+
+def get_spatial_coverage_attributes(ds,
+                                    time='time',
+                                    lat='latitude',
+                                    lon='longitude',
+                                    depth='depth',
+                                    ):
+    """
+    This method generates the geospatial and time coverage attributes associated to an xarray dataset.
+    """
+    time_spatial_coverage = {}
+    # time
+    if time in ds:
+        time_spatial_coverage.update({
+            'time_coverage_start': str(ds[time].min().values),
+            'time_coverage_end': str(ds[time].max().values),
+            'time_coverage_duration': str((ds[time].max() - ds[time].min())
```

Comment:
Does this have to be in duration format? see https://wiki.esipfed.org/Attribute_Convention_for_Data_Discovery_1-3#time_coverage_duration

Reply:
This was changed to export to isoformat with pandas.

Reply:
oh good find